### PR TITLE
emit export declarations for system modules as a part of 'execute' me…

### LINKED ISSUE #4162: Variables are not exported properlly when using ES6 Modules with default export and targetting SystemJS.
There is problem when I try to re export some module imported from export default.

Link to gist: https://gist.github.com/majo44/4d84feac30ad5a803658
Link to working example (copied compiled code): https://jsfiddle.net/fjfppk0z/

### PR DESCRIPTION
**Pull request by vladima**
…thod. fixes #4162. New output is aligned with what Babel will emit in the similar case

### CODE REVIEW

**Comment by vladima** [comment]
pinging @mhegazy

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -3090,32 +3090,39 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                        emitEnd(specifier.name);
+                        write(" = ");
+                        emitExpressionIdentifier(name);
                         write(";");
                     }
                 }
             }
+            
+            function emitExportSpecifierInSystemModule(specifier: ExportSpecifier): void {
+                Debug.assert(compilerOptions.module === ModuleKind.System);
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: @sheetalkamat this looks related to your other change.. can u take a look.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -6226,26 +6201,41 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
                     switch (statement.kind) {
+                        // - function declarations are not emitted because they were already hoisted
+                        // - import declarations are not emitted since they are already handled in setters
+                        // - export declarations with module specifiers are not emitted since they were already written in setters
+                        // - export declarations without module specifiers are emitted preserving the order
+                        case SyntaxKind.FunctionDeclaration:
                         case SyntaxKind.ExportDeclaration:
                         case SyntaxKind.ImportDeclaration:
-                        case SyntaxKind.FunctionDeclaration:
+                            if (statement.kind === SyntaxKind.ExportDeclaration) {
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: move this to a different case label instead of checking the kind again.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -6105,115 +6112,83 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
-                                }
-                                else {
-                                    // emit re-exports for named imports
-                                    // import {a, b} from 'foo'
-                                    // export {a, b as c}
-                                    for (let element of (<NamedImports>namedBindings).elements) {
-                                        emitExportMemberAssignments(element.name || element.propertyName);
+                                    write(`var ${reexportsVariableName} = {};`);
+                                    writeLine();
+                                    for (let e of (<ExportDeclaration>entry).exportClause.elements) {
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: i would replace this with:
```ts
for (let element of (<NamedImports>namedBindings).elements) {
    emitExportMemberAssignments(element.name || element.propertyName);
    //write(`var ${reexportsVariableName} = {};`);
    write(`${exportFunctionForFile}({`);
    increaseIndent();
    writeLine();
    for (let e of (<ExportDeclaration>entry).exportClause.elements) {
        write(`"`);
        emitNodeWithoutSourceMap(e.name);
        write(`": ${parameterName}["`);
        emitNodeWithoutSourceMap(e.propertyName || e.name);
        write(`"];`);
        writeLine();
    }
    decreaseIndent();
    write(`});`);
}
```
saves you the extra variable declaration and one line less :D

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -3090,32 +3090,39 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 }
             }
 
-            function emitExportMemberAssignments(name: Identifier) {
+            function emitExportMemberAssignmentsInNonSystemModule(name: Identifier) {
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: why not just keep this `emitExportMemberAssignments` eith a return at the top and note that it is handled in System emit?

File: src/compiler/emitter.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/emitter.ts]: since this is just a matter of name, I can roll it back.

File: src/compiler/emitter.ts (hunk 3 above)
Comment (vladima) [inline src/compiler/emitter.ts]: ok

**Comment by sheetalkamat** [comment]
Lgtm. Once you checkin I will rebase my changes of your changes

**Pull request merged**